File: ant-sdk/antd-swift/Sources/AntdSdk/AntdClient.swift
import Foundation

/// Factory for creating Autonomi SDK clients.
///
/// ```swift
/// let client = AntdClient.createRest()
/// let health = try await client.health()
/// ```
public enum AntdClient {

    /// Create a REST client connecting to the antd daemon.
    /// - Parameters:
    ///   - baseURL: Base URL of the antd REST API (default: http://localhost:8082)
    ///   - timeout: Request timeout in seconds (default: 300)
    public static func createRest(
        baseURL: String = "http://localhost:8082",
        timeout: TimeInterval = 300
    ) -> AntdClientProtocol {
        AntdRestClient(baseURL: baseURL, timeout: timeout)
    }

    /// Create a gRPC client connecting to the antd daemon.
    ///
    /// > Requires macOS 15+ / iOS 18+ / tvOS 18+ / watchOS 11+ (grpc-swift 2.x).
    /// > Use ``createRest(baseURL:timeout:)`` on older platforms.
    ///
    /// - Parameter target: gRPC target address (default: localhost:50051)
    @available(macOS 15.0, iOS 18.0, tvOS 18.0, watchOS 11.0, *)
    public static func createGrpc(target: String = "localhost:50051") -> AntdClientProtocol {
        AntdGrpcClient(target: target)
    }

    /// Create a client using the specified transport.
    ///
    /// > `transport: "grpc"` requires macOS 15+ / iOS 18+ / tvOS 18+ / watchOS 11+.
    ///
    /// - Parameters:
    ///   - transport: "rest" or "grpc"
    ///   - endpoint: Optional custom endpoint override
    public static func create(transport: String = "rest", endpoint: String? = nil) -> AntdClientProtocol {
        switch transport.lowercased() {
        case "rest":
            return createRest(baseURL: endpoint ?? "http://localhost:8082")
        case "grpc":
            if #available(macOS 15.0, iOS 18.0, tvOS 18.0, watchOS 11.0, *) {
                return createGrpc(target: endpoint ?? "localhost:50051")
            } else {
                fatalError("gRPC transport requires macOS 15+ / iOS 18+. Use 'rest' on older platforms.")
            }
        default:
            fatalError("Unknown transport: \(transport). Use 'rest' or 'grpc'.")
        }
    }
}

File: ant-sdk/antd-kotlin/lib/src/main/kotlin/com/autonomi/sdk/AntdClient.kt
package com.autonomi.sdk

import java.time.Duration

/**
 * Factory for creating Autonomi SDK clients.
 *
 * ```kotlin
 * val client = AntdClient.createRest()
 * val health = client.health()
 * ```
 */
object AntdClient {

    /**
     * Create a REST client connecting to the antd daemon.
     * @param baseUrl Base URL of the antd REST API (default: http://localhost:8082)
     * @param timeout Request timeout (default: 300 seconds)
     */
    @JvmStatic
    @JvmOverloads
    fun createRest(
        baseUrl: String = "http://localhost:8082",
        timeout: Duration = Duration.ofSeconds(300),
    ): IAntdClient = AntdRestClient(baseUrl, timeout)

    /**
     * Create a gRPC client connecting to the antd daemon.
     * @param target gRPC target address (default: localhost:50051)
     */
    @JvmStatic
    @JvmOverloads
    fun createGrpc(target: String = "localhost:50051"): IAntdClient = AntdGrpcClient(target)

    /**
     * Create a client using the specified transport.
     * @param transport "rest" or "grpc"
     * @param endpoint Optional custom endpoint override
     */
    @JvmStatic
    @JvmOverloads
    fun create(transport: String = "rest", endpoint: String? = null): IAntdClient =
        when (transport.lowercase()) {
            "rest" -> createRest(endpoint ?: "http://localhost:8082")
            "grpc" -> createGrpc(endpoint ?: "localhost:50051")
            else -> throw IllegalArgumentException("Unknown transport: $transport. Use 'rest' or 'grpc'.")
        }
}
